cdk/src/open-data-platform/data-plane/data-import/handler-factory.ts

```diff
@@ -110,106 +112,80 @@ const readGeoJsonFile = (
 ): Promise<ProcessResult> =>
   new Promise((resolve, reject) => {
     let batchId = 0; // Not-necessarily-sequential ID used to group logs.
-    let processedRowCount = 0; // Number of rows processed, either successfully or with errors.
-    let skippedRowCount = 0; // Number of rows skipped due to offset.
-    let inProcessRowCount = 0; // Number of rows currently being proccessed.
+    let savedRowCount = 0;
+    let failedRowCount = 0;
+    let inProcessRowCount = 0;
+    const failedBatches: Batch[] = [];
+
+    interface Batch {
+      id: string;
+      rows: any[];
+    }
 
     console.log('Starting to process file in S3:', s3Params);
 
-    // Store all of the promises returned by the callback so the Lambda waits for their resolution
-    // before exiting. They will start to execute before they are awaited and run asynchonously
-    // in-parallel with no guarantee of order.
-    const promises: Promise<any>[] = [];
+    const handleBadBatch = (e: Error, batch: Batch) => {
+      console.log(`${batch.id} failed. Scheduling retry...`, e);
+      failedBatches.push(batch);
+    };
+
+    const saveBatch = async (batch: Batch) => {
+      const descriptor = `${batch.id} of ${batch.rows.length} rows`;
+      console.log(`processing ${descriptor}`);
+      const start = Date.now();
+      try {
+        inProcessRowCount += batch.rows.length;
+        await callback(batch.rows, db);
+        savedRowCount += batch.rows.length;
+      } catch(e) {
+        inProcessRowCount -= batch.rows.length;
+        throw e; // handle in handleBadBatch
+      }
+      console.log(`${descriptor} saved in ${Date.now() - start} ms`);
+    };
+
+    // Called once the pipeline finishes and asks the sink to flush the last few rows.
+    // Not called if the source dries up and the sink has nothing to flush. 
```

Comment:
I *could* mean that. `handleEndOfFileStream` is used in two places:
- parallel-sink
  - If `flush` is called upstream, we invoke it 
  - Also, if the underlying stream errors out, we invoke it using the [stream finish handler](https://nodejs.org/api/stream.html#event-finish)
  - If there's nothing to flush and the underlying stream hasn't errored, then parallel sink sends a 'finish' event.
- We catch that `finish` event on line 187 (`pipeline.on('finish', handleEndOfFilestream);`

This *should* ensure it's always called. 

One thing I realize is that it *might* be called more than once.
- I haven't managed to make that happen though, so not sure how much to invest it fixing it
- We can explicitly prevent it by using `_.once()` or similar, but I didn't want to make this more complicated than it already is or add more external libraries without a good reason. Happy to do this if you think it's a good idea though.

Reply:
Nah, I don't think we need to prevent double-execution. If some rows are imported twice, the `ON CONFLICT` clause should make sure the 2nd is ignored.

Reply:
But maybe edit this comment, since "not called if..." made me think there was a possibility that `handleEndOfFilestream` won't get called (and therefore `failedBatches` never gets a retry)